### PR TITLE
Fixes here and there for s7 scheme on Windows 10

### PR DESCRIPTION
**Pull request by 326623**
1. Memory tuning for s7 (please take a look at the comments and the scheme code I added).
2. I wasn't able to build with leaving Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin) in the code there. And I don't know how to resolve it for msys2-ucrt build.
3. I think we can even change the code in file.cpp that relies on unix api to Qt5 on Windows; we already depend on it anyway and it might even allow us to build TeXmacs using visual studio's toolchain, which has various handy profilers that can help us evaluate performance and pinpoint memory issues.

### CODE REVIEW

**Comment by darcy-shen** [comment]
Awesome work! I do not have much time to review your pull request (I mean for windows related changes, I have to test it on Windows).

Next weekends, I will try to merge part (`get_user_login`, `get_user_name`) of the improvements into the master branch by @mgubi and then I will review your changes (might be two weeks later).

To integrate S7, we need to merge the changes commits by commits. First of all, the common changes (not binding to S7) will be merged.

File: src/Scheme/S7/s7.c [hunk 1]
```diff
@@ -7208,8 +7208,17 @@ static void resize_heap_to(s7_scheme *sc, int64_t size)
       /* (sc->heap_size < 2048000) */  /* 8192000 here improves various gc benchmarks only slightly */
       /* maybe the choice of 4 should depend on how much space was freed rather than the current heap_size? */
       if (old_free < old_size * sc->gc_resize_heap_by_4_fraction)
-	sc->heap_size *= 4;          /* *8 if < 1M (or whatever) doesn't make much difference */
-      else sc->heap_size *= 2;
+	sc->heap_size *= 2;          /* *8 if < 1M (or whatever) doesn't make much difference */
```
Comment (darcy-shen) [inline src/Scheme/S7/s7.c:7211]: this part can be submitted to S7 like providing S7 internal options to control the behavior

File: src/Plugins/Windows/nowide/cstdio.hpp [hunk 2]
```diff
@@ -144,22 +144,24 @@ inline DIR* opendir(char const *name)
     return _wopendir (wname.c_str());
 }
 
-inline char* readir_entry(DIR* dir)
```
Comment (darcy-shen) [inline src/Plugins/Windows/nowide/cstdio.hpp:147]: Plugins/Windows/nowide is imported from Boost Nowide

It is not a good practice to modify the Nowide library directory. We could write a TeXmacs internal readir_entry somewhere else.

**Comment by mgubi** [comment]
I would like to propose to move the development of the s7 support to the wip_s7 branch I've just created in the repo. I find confusing to have multiple branches s7_branch_2.1.1 and s7_branch_2.1 . I think that in this development phase we should keep s7 integrate to the last revision of the master branch. That is why I propose to move all to wip_s7. I will discontinue my personal branch in mgubi/texmacs/s7 and move my work here. @darcy-shen  @326623 let me know if this is ok for you. Also when modifying s7.{c,h} please keep the changes also in a patch file, since from time to time I update s7 to its last version and if we do not keep track of our local changes they will be lost. Ideally I would like to keep the changes to the source code at the bare minimum and ask to implement them upstream if they are really needed. Changes to the memory management are delicate, so please lets also discuss them in a separate PR on wip_s7 instead that bundling them in a PR like this one which has a lot of other code.

File: TeXmacs/progs/init-texmacs-s7.scm [hunk 3]
```diff
@@ -529,3 +542,8 @@
                   ;(quit-TeXmacs)
                   ))))))))))))
 
+;; iota is not implemented for s7.
```
Comment (mgubi) [inline TeXmacs/progs/init-texmacs-s7.scm:545]: Do not use tm-define for simple general utility, tm-define has some overhead and is not needed here if you do not plan to leave to the user the possibility to override this functionality dynamically. Look at how the compatibility layer is implemented in (kernel boot compat-s7) and put this function there, as a public define.

File: src/Plugins/Qt/qt_gui.cpp [hunk 4]
```diff
@@ -875,7 +878,7 @@ qt_gui_rep::update () {
   
   time_t delay = delayed_commands.lapse - texmacs_time();
   if (needing_update) delay = 0;
-  else                delay = max (0, min (std_delay, delay));
+  else                delay = max (0, min (std_delay, static_cast<int>(delay)));
```
[the remark this answers is not in the document]
Comment (326623) [reply src/Plugins/Qt/qt_gui.cpp:881]: It fails to compile on my system; my newer compiler is complaining about ambiguity of function overloading.

**Comment by 326623** [comment]
Hi Max, I suppose you have integrated all the ideas in this pull request in the latest s7 dev branch? If so, I will proceed to close this pull request and work on the latest s7 branch? Thanks for all the comments!

**Comment by mgubi** [comment]
I believe I did. I've however integrated only the code relative to S7 and left away the changes to C++ relative to Windows. Thanks for the feedback! and please continue testing the S7 part, in particular I would like to see if we can reduce the memory footprint.

**Pull request closed without merging**